controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,260 @@
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect requested constraints and add found kinds which is used in contraint to config.spec.sync.syncOnly
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+
+	// Get config or create if not exist
+	config := &v1alpha1.Config{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      "config",
+	}, config)
+
+	if apierrors.IsNotFound(err) {
+		config = &v1alpha1.Config{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "config",
+				Namespace: r.Namespace,
+			},
+		}
+		err = r.Create(ctx, config)
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	contraintPodStatus := &v1beta1.ConstraintPodStatus{}
+	err = r.Get(ctx, request.NamespacedName, contraintPodStatus)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(err, "Cannot find any contraintStatus")
+
+			return reconcile.Result{}, nil
+		}
+		// Requeue
+		return reconcile.Result{}, err
+	}
+
+	labels := contraintPodStatus.GetLabels()
+
+	constraintKind := labels["internal.gatekeeper.sh/constraint-kind"]
+	constraintName := labels["internal.gatekeeper.sh/constraint-name"]
+
+	table := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for uniqueue filtering
+	for _, entry := range config.Spec.Sync.SyncOnly {
+		table[entry] = true
+	}
+
+	constraintGVR := schema.GroupVersionResource{
+		Group:    "constraints.gatekeeper.sh",
+		Version:  "v1beta1",
+		Resource: strings.ToLower(constraintKind),
+	}
+
+	constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, constraintName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	contraintSyncOnlyEntries, err := r.getSyncOnlys(constraintMatchKinds)
+	if err != nil {
+		log.Error(err, "Error to get matching kind and apigroup")
+
+		return reconcile.Result{}, nil
+	}
+
+	for _, ce := range *contraintSyncOnlyEntries {
+		table[ce] = true
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+	for key := range table {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	config.Spec.Sync.SyncOnly = syncOnlys
+	err = r.Update(ctx, config, &client.UpdateOptions{})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ConstraintStatusReconciler) getSyncOnlys(constraintMatchKinds []interface{}) (
+	*[]v1alpha1.SyncOnlyEntry, error,
+) {
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for _, kind := range constraintMatchKinds {
+		newKind := kind.(map[string]interface{})
```

Comment:
Is it safe to do type assertions without checking? What if one of the keys isn't set in `kind`?